components/dashboard/src/settings/PersonalAccessTokens.tsx

```diff
@@ -214,153 +304,199 @@ export function PersonalAccessTokenCreateView() {
+                                            checked={item.scopes.every((s) => token.scopes.has(s))}
+                                            onChange={(e) => {
+                                                if (e.target.checked) {
+                                                    update({}, item.scopes);
+                                                } else {
+                                                    update({}, undefined, item.scopes);
+                                                }
+                                            }}
+                                        />
+                                    ))}
+                                </div>
+                            </div>
                         </div>
                     </div>
-                </div>
-                <div className="flex gap-2">
-                    {editTokenID && (
-                        <Link to={settingsPathPersonalAccessTokens}>
-                            <button className="secondary" onClick={handleConfirm}>
-                                Cancel
-                            </button>
-                        </Link>
-                    )}
-                    <button onClick={handleConfirm}>{editTokenID ? "Update" : "Create"} Personal Access Token</button>
-                </div>
+                    <div className="flex gap-2">
+                        {editTokenID && (
+                            <Link to={settingsPathPersonalAccessTokens}>
+                                <button className="secondary" onClick={handleConfirm}>
+                                    Cancel
+                                </button>
+                            </Link>
+                        )}
+                        <button onClick={handleConfirm}>
+                            {editTokenID ? "Update" : "Create"} Personal Access Token
+                        </button>
+                    </div>
+                </SpinnerOverlayLoader>
             </PageWithSettingsSubMenu>
         </div>
     );
 }
 
+interface TokenEntryProps {
+    token: PersonalAccessToken;
+    menuEntries: ContextMenuEntry[];
+}
+
+function TokenEntry(props: TokenEntryProps) {
+    const [dateInfo, setDateInfo] = useState<{ expired: boolean; content: string }>();
+
+    useEffect(() => {
+        if (!props.token.expirationTime) {
+            setDateInfo({ expired: false, content: "" });
+        }
+        const tmp = dayjs(props.token.expirationTime!.toDate());
+        setDateInfo({ expired: tmp.isBefore(dayjs()), content: tmp.format("MMM D, YYYY") });
+    }, [props.token.expirationTime]);
+
+    const getScopes = () => {
+        if (!props.token.scopes) {
+            return "";
+        }
+        const permissions = AllPermissions.filter((e) => e.scopes.every((v) => props.token.scopes.includes(v)));
+        if (permissions.length > 0) {
+            return permissions.map((e) => e.name).join("\n");
+        } else {
+            return "No access";
```

Comment:
**question**: @jeanp413 I remember briefly discussing about this, but does it make sense to allow tokens to be generated without any access? Cc @easyCZ

Reply:
Yes, absolutely. 

Ideally, you generate a token with no access at all to begin with. You use it in your code and get back an error which says "you're missing permission X". You then go and add exactly that permission. 

Yes, currently we have tokens which only have all access. But even for these there's value in tokens with no perms. For example you suspect it leaked, but need to investigate. It's better to remove all perms rather than delete it immediately in case it did not leak, or you do not need to regenerate.

Reply:
Thanks so much for clarifying, @easyCZ! 🌮 🌮